dashboard/app/models/regional_partner.rb

```diff
@@ -58,16 +58,24 @@ def program_manager=(program_manager_id)
   # @param [String] zip_code
   # @param [String] state - 2-letter state code
   def self.find_by_region(zip_code, state)
-    return RegionalPartner.
-      joins(:mappings).
-      where(pd_regional_partner_mappings: {state: state}).
-      or(
-        joins(:mappings).
-        where(pd_regional_partner_mappings: {zip_code: zip_code})
-      ).
-      # prefer match by zip code when multiple partners cover the same state
-      order('pd_regional_partner_mappings.zip_code IS NOT NULL DESC').
-      first
+    return nil if zip_code.nil? && state.nil?
+
+    find_by_region_query = RegionalPartner
+    if state
+      find_by_region_query = find_by_region_query.joins(:mappings).where(pd_regional_partner_mappings: {state: state})
+      if zip_code
+        find_by_region_query = find_by_region_query.or(
+          joins(:mappings).
+          where(pd_regional_partner_mappings: {zip_code: zip_code})
+        )
+      end
+    elsif zip_code
+      find_by_region_query = find_by_region_query.
+        joins(:mappings).where(pd_regional_partner_mappings: {zip_code: zip_code})
+    end
```

Comment:
This looks a bit cumbersome. The `joins(:mappings)` is repeated in all cases, and the zip_code query is repeated. 

How about something like this isntead?
```ruby
base_query = RegionalPartner.joins(:mappings)
queries = []
queries << base_query.where(pd_regional_partner_mappings: {state: state}) if state
queries << base_query.where(pd_regional_partner_mappings: {zip_code: zip_code}) if zip_code

full_query = queries.length == 1 ? queries[0] : queries[0].or(queries[1])
```

Reply:
Yup.  I found a variation on that approach that works.